cpp/src/arrow/compute/kernels/scalar_string_ascii.cc

```diff
@@ -2347,6 +2351,136 @@ void AddAsciiStringExtractRegex(FunctionRegistry* registry) {
   }
   DCHECK_OK(registry->AddFunction(std::move(func)));
 }
+struct ExtractRegexSpanData : public BaseExtractRegexData {
+  static Result<ExtractRegexSpanData> Make(const std::string& pattern,
+                                           bool is_utf8 = true) {
+    auto data = ExtractRegexSpanData(pattern, is_utf8);
+    ARROW_RETURN_NOT_OK(data.Init());
+    return data;
+  }
+
+  Result<TypeHolder> ResolveOutputType(const std::vector<TypeHolder>& types) const {
+    const DataType* input_type = types[0].type;
+    if (input_type == nullptr) {
+      return nullptr;
+    }
+    DCHECK(is_base_binary_like(input_type->id()));
+    const size_t field_count = num_groups();
+    FieldVector fields;
+    fields.reserve(field_count);
+    const auto owned_type = input_type->GetSharedPtr();
+    auto type = is_binary_like(owned_type->id()) ? int32() : int64();
+    for (const auto& group_name : group_names) {
+      // size list is 2 as every span contains position and length
+      fields.push_back(field(group_name, fixed_size_list(type, 2)));
+    }
+    return struct_(fields);
```

Comment:
Can move the vector to avoid a temporary copy
```suggestion
    return struct_(std::move(fields));
```